### ISSUE TITLE
AuthorizationServerSecurityConfiguration should allow overriding oauth endpoint authorization

### ISSUE THREAD

**Issue by btiernay** [spring-projects/spring-security-oauth#496]
Currently the `protected void configure(HttpSecurity http) throws Exception ` method in `AuthorizationServerSecurityConfiguration` defines the authorization of the oauth endpoints, making it very difficult (impossible?) to change:

```java
		String tokenEndpointPath = handlerMapping.getServletPath("/oauth/token");
		String tokenKeyPath = handlerMapping.getServletPath("/oauth/token_key");
		String checkTokenPath = handlerMapping.getServletPath("/oauth/check_token");
...
		http
        	.authorizeRequests()
            	.antMatchers(tokenEndpointPath).fullyAuthenticated() // Cannot change
            	.antMatchers(tokenKeyPath).access(configurer.getTokenKeyAccess())
            	.antMatchers(checkTokenPath).access(configurer.getCheckTokenAccess())
        .and()
        	.requestMatchers()
            	.antMatchers(tokenEndpointPath, tokenKeyPath, checkTokenPath);
```

It is unclear how to define fine grained authorization on endpoints such as `/oauth/token`. Please see this question for more details:

http://stackoverflow.com/questions/30488304/contolling-access-to-spring-security-oauth2-endpoints

In the documentation it is written:

> N.B. the Authorization endpoint /oauth/authorize (or its mapped alternative) should be protected using Spring Security so that it is only accessible to authenticated users. For instance using a standard Spring Security WebSecurityConfigurer:

```java
   @Override
    protected void configure(HttpSecurity http) throws Exception {
        http
            .authorizeRequests().antMatchers("/login").permitAll().and()
        // default protection for all resources (including /oauth/authorize)
            .authorizeRequests()
                .anyRequest().hasRole("USER")
        // ... more configuration, e.g. for form login
    }
```

However, since the `authorizeRequests` call in `AuthorizationServerSecurityConfiguration` wipes out this configuration, it does not seem possible.

**Comment by dsyer** [comment]
`AuthorizationServerSecurityConfiguration` creates a filter with `order=0`. To secure `/oauth/authorize` you only have to add one with a lower order. So do you need to configure fine-grained security fro `/oauth/token` or not (you didn't really explain why)?

**Comment by btiernay** [comment]
@dsyer our use case is:

- allow only certain clients to generate tokens (`/oauth/token`), but
- allow all clients to validate tokens (`/oauth/authorize`)

We are stuck trying to configure the first bit. Is this use case possible currently?

Much thanks.

**Comment by dsyer** [comment]
You have clients that are not allowed to use the `/token` endpoint? That's unusual. Are they all only allowed "implicit" grant type or something?

**Comment by btiernay** [comment]
No, its just that we want to harden the endpoint to prevent certain clients from creating tokens. I realize this is a bit unusual, but I still think this is a valid use case. In general, you might want to filter by IP, role, etc., as well.

**Comment by dsyer** [comment]
As far as I know there's nothing stopping you from adding a security filter upstream of the default one and weeding out any requests you don't like. That's the beauty of having the /token endpoint as a plain MVC feature (it's orthogonal to Security). I haven't tried it, but I bet you can do it with a `WebSecurityAdapter` with high order and a request pattern `/oauth/token/**`.

**Comment by dl1ely** [comment]
I have to jump on that issue in a different context. I need to get hold of the `RequestCacheConfigurer` for the `/oauth/` endpoints. I would normally do that by calling `HttpSecurity.requestCache()` in the `configure` method of `SecurityConfigurerAdapter`. All the configuration is done in `AuthorizationServerSecurityConfigurer.configure(HttpSecurity http)`, but i do not see how to hook in there and provide `HttpSecurity` with my `RequestCache` implementation via the `RequestCacheConfigurer`. Any hints on that?

**Comment by nwallman** [comment]
I wanted to add that I believe this should be customizable to where we can bypass authentication when attempting to create an access token.  Unless I've mis-read the spec, the client_id and client_secret are optional for a resource owner password grant flow which is precisely what I'm trying to implement.  In this case my client is a mobile device and therefore I can't trust the client and the client secret isn't going to do me much good. 

Also @dsyer I tried your approach of building a customer WebSecurityConfig and giving it a lower order and it appeared to cause additional issues so I abandoned that.  I really don't see why we can't just allow the tokenEndpointPath to be customizable long with the tokenKeyPath and checkTokenPath.

OAuth 2.0 spec: https://tools.ietf.org/html/rfc6749#section-4.3
